### PR TITLE
Fix shutdown SIGSEGV of Channelz Registry with gevent

### PR DESCRIPTION
**Pull request by lidizheng**
Fixes https://github.com/grpc/grpc/issues/20299

---

The `Delete` in our `memory.h` seems not setting pointer back to `nulltpr`. The deletion seems half-done. I don't have a comprehensive fix for it, but with several lines of changes the segfault of gevent should go away.

### CODE REVIEW

**Comment by gnossen** [comment]
This appears to be a bandaid. The number of increfs should equal the number of decrefs, right? Otherwise, the reference count is not actually accurate. Can you summarize your findings on the chain of events that result in a destructed pointer getting decref'd again?

**Comment by lidizheng** [comment]
`ChannelzRegistry` is a singleton that got assign to a global variable, and it is not a ref counted instance. So... the `Delete` does mean delete the instance. @yashykt Should the `Delete` call set the caller pointer to `nullptr`?

My current finding can be found in https://github.com/grpc/grpc/issues/20299#issuecomment-533670570. In short, the shutdown process inherently has a data race that only reveals when there is only one thread (custom IO manager enabled).

I still didn't figure out how the shutdown and dealloc got ordered across threads. If you can help digging, it would be great!

**Comment by markdroth** [comment]
This does not look like the right fix.

`Delete()` is not supposed to set the pointer to null.  And the channelz `Shutdown()` should always be called before it is re-`Init()`ed.

**Comment by lidizheng** [comment]
@markdroth I think I need help with the shutdown path in Channelz. Can you point a direction that you think could be the right solution?

For current solution, my fix doesn't affect the next `Init()` invocation, since it will assign the address of the new instance of `ChannelzRegistry`.

---

Pasting the weird shutdown order:

The normal call order is
* `ChannelzRegistry::InternalUnregister`
* `grpc_shutdown_blocking`
* `ChannelzRegistry::InternalUnregister` <- Happen in another thread
* `ChannelzRegistry::Shutdown` <-Deleted the lock

But in the gevent version, the order becomes:
* `ChannelzRegistry::InternalUnregister`
* `grpc_shutdown_blocking`
* `ChannelzRegistry::Shutdown` <-Deleted the lock
* `ChannelzRegistry::InternalUnregister` <- Abort

**Comment by markdroth** [comment]
Channelz nodes unregister themselves when they are destroyed.  It sounds to me like you have a channelz node somewhere that is outliving the channelz registry.  You should figure out which one and make sure that the node is destroyed before the registry is shut down.

**Comment by yashykt** [comment]
Is this ready for review?

**Comment by lidizheng** [comment]
@yashykt Sorry for spamming, @AspirinSJL just remind me the existence of this PR. I wasn't found enough time to dig into the behavior one month ago, I would probably try to find sometime this week to investigate the root cause.

**Comment by yashykt** [comment]
Forwarding Juanli's suggestion that if you actively developing a PR, consider doing it in a draft PR mode :)

**Comment by lidizheng** [comment]
I will close this one for now. @yashykt
